### PR TITLE
Tiny Instruction Cache

### PR DESCRIPTION
**Pull request by sripathi-muralitharan**
This PR adds support for a tiny cache - 4KB and a corresponding tiny config (reflects only the tiny icache changes) that is supported only in the UCE. 

A similar change for the D$ will soon follow in a future PR.

### CODE REVIEW

**Comment by taylor-bsg** [comment]
Nice work!

This is a lot of extra code to support a corner case.  Lot of it has to do with signal width underflow kind of conditions. Unfortunately, this means that meta programming is only partially working, and our code gets multiplied up exponentially with all of these cases.

My suggestion is that as much as possible you should try to push the underflow down into the leaf modules, so that they are more robust in general to this sort of thing, for example the LRU module or the stat ram. Then when we write other code that has underflow, it will just work out of the box. A simple example of this is the variable #input mux module, and making it handle the case of only one incoming item. In general we have created the macro BSG_SAFE_CLOG2 for underflow cases where only one item is there, and the CLOG2 function returns 0, results in a zero width signal. In this case we may need something like BSG_SAFE_MINUS.

M

**Comment by dpetrisko** [comment]
^ took the words right out of my mouth

**Comment by sripathi-muralitharan** [comment]
Opened a PR in Basejump-STL with this macro. PR 352 - https://github.com/bespoke-silicon-group/basejump_stl/pull/352. I hope this is sufficient to solve the meta programming problem here. 

I haven't made changes to the code in BP yet (Waiting for the one in Basejump-STL to be merged in)

**Comment by dpetrisko** [comment]
You should make the changes, point the submodule to your basejump stl branch, and then write “depends on #xyz” as a comment here. That’s the best way to track these things

File: bp_fe/src/v/bp_fe_icache.v
Comment (dpetrisko) [inline bp_fe/src/v/bp_fe_icache.v:101]: do the max(2) in the macro

File: bp_fe/src/v/bp_fe_icache.v
Comment (dpetrisko) [inline bp_fe/src/v/bp_fe_icache.v:32]: max(2) in macro

File: bp_fe/src/v/bp_fe_icache.v
Comment (dpetrisko) [inline bp_fe/src/v/bp_fe_icache.v:42]: This comment is confusing -- instead, I would draw an ascii diagram

File: bp_fe/src/v/bp_fe_icache.v
Comment (dpetrisko) [inline bp_fe/src/v/bp_fe_icache.v:44]: having both bank_offset_width and mem_bank_offset_width is very confusing.  Why not just set bank_offset_width correctly?

File: bp_fe/src/v/bp_fe_icache.v
Comment (dpetrisko) [inline bp_fe/src/v/bp_fe_icache.v:314]: These comments about direct-mapped cache are unnecessary.  It's a very niche configuration

File: bp_fe/src/v/bp_fe_icache.v
Comment (dpetrisko) [inline bp_fe/src/v/bp_fe_icache.v:300]: @taylor-bsg Crazy idea, but does it make sense to support 0-width memories as well?

**Comment by dpetrisko** [comment]
@taylor-bsg .  Most of the other corner cases are related to 0-width concatenations e.g. {index, offset}, where offset is 0 bits. I can't think of a good way to get around that

**Comment by taylor-bsg** [comment]
Yeah I suggested to Sripathi that we support 0-width memories rather than a
separate generate statement.

BSG_SAFE_CONCAT ?



M

**Comment by dpetrisko** [comment]
> BSG_SAFE_CONCAT

Hm, we need to signal to the macro if the field is actually a 0-width signal disguised as a 1-width signal `BSG_SAFE_CONCAT(index, offset) makes it ambiguous if offset is actually 0 or 1 bits.

`BSG_SAFE_CONCAT(index, offset, $clog2(offset)) would solve this problem but is a little messy.

File: bp_fe/src/v/bp_fe_icache.v
Comment (sripathi-muralitharan) [reply bp_fe/src/v/bp_fe_icache.v:44]: That's because when defining widths, I wanted them to be one so that I don't get bit vectors with negative indices. But for places where I want to index using the +:, I wanted a zero to be correct.

**Comment by sripathi-muralitharan** [comment]
For BSG_SAFE_CONCAT, is it okay to enforce ordering in the parameters? Like the first parameter has to be something non-zero and the second one could be 0 or non-zero?

File: bp_fe/src/v/bp_fe_icache.v
Comment (sripathi-muralitharan) [reply bp_fe/src/v/bp_fe_icache.v]: Got rid of one of them and reverted back to the previous state of only 1 offset parameter by leveraging BSG_SAFE_MINUS

**Comment by sripathi-muralitharan** [comment]
Managed to solve the other 0-width concatenations issue by using a conditional concatenation feature in Verilog. Ready for review again.

File: bp_me/src/v/cce/bp_uce.v
Comment (dpetrisko) [inline bp_me/src/v/cce/bp_uce.v:298]: I think you can use the zero width concatentation here

File: bp_me/src/v/cce/bp_uce.v
Comment (dpetrisko) [inline bp_me/src/v/cce/bp_uce.v:305]: Do you need the max here? If fill_size_in_bank == 1 for direct mapped, then this clause will never get called

File: bp_me/src/v/cce/bp_uce.v
Comment (dpetrisko) [inline bp_me/src/v/cce/bp_uce.v:291]: Is bank_width == block_width? I thought there was just an inequality there

File: bp_me/src/v/cce/bp_uce.v
Comment (dpetrisko) [inline bp_me/src/v/cce/bp_uce.v:26]: If you keep this as SAFE_CLOG2 and use zero width concatentation, I believe that will get rid of all the edge cases below

File: bp_fe/src/v/bp_fe_top.v
Comment (dpetrisko) [inline bp_fe/src/v/bp_fe_top.v:20]: I don't think any of these parameters are used?

File: bp_fe/src/v/bp_fe_icache.v
Comment (dpetrisko) [inline bp_fe/src/v/bp_fe_icache.v:641]: can just use '0 and '1

File: bp_fe/src/v/bp_fe_icache.v
Comment (dpetrisko) [inline bp_fe/src/v/bp_fe_icache.v:614]: bsg_safe_minus?

File: bp_fe/src/v/bp_fe_icache.v
Comment (dpetrisko) [inline bp_fe/src/v/bp_fe_icache.v:512]: zero width concat

File: bp_fe/src/v/bp_fe_icache.v
Comment (dpetrisko) [inline bp_fe/src/v/bp_fe_icache.v:376]: goes away with zero width memory

File: bp_common/src/include/bp_common_cache_engine_if.vh
Comment (dpetrisko) [inline bp_common/src/include/bp_common_cache_engine_if.vh:159]: safe minus

File: bp_me/src/v/cce/bp_uce.v
Comment (sripathi-muralitharan) [reply bp_me/src/v/cce/bp_uce.v:291]: The general inequality is bank_width <= block_width. For the direct-mapped case, it becomes bank_width = block_width

File: bp_fe/src/v/bp_fe_top.v
Comment (sripathi-muralitharan) [reply bp_fe/src/v/bp_fe_top.v:20]: Yes, there are some redundant parameters. I can get rid of them.

File: bp_fe/src/v/bp_fe_icache.v
Comment (sripathi-muralitharan) [reply bp_fe/src/v/bp_fe_icache.v:376]: Not so sure about this. Seems to me that the priority encoder from where the `invalid_exist` and `way_invalid_index` are generated end up throwing out X's when width_p = 1. Seems to me that these corner conditions haven't been handled fully/correctly. This means, that will end up putting an X on the metadata way. I am guessing it may not become a problem since its all direct mapped and there is no concept of a way, but I am worried about the X-propagation that might cause issues down the line? What do you think about this?

File: bp_me/src/v/cce/bp_uce.v
Comment (sripathi-muralitharan) [reply bp_me/src/v/cce/bp_uce.v:26]: I believe that will not calculate some of the other parameters correctly. So, I think its best to keep it the way it is. But exactly which edge cases are you talking about? The ones with the indexed-part selects (eg. line 305 in `bp_uce.v`)?

File: bp_me/src/v/cce/bp_uce.v
Comment (sripathi-muralitharan) [reply bp_me/src/v/cce/bp_uce.v]: Yup. Fixed that in the latest commit.

File: bp_fe/src/v/bp_fe_icache.v
Comment (dpetrisko) [reply bp_fe/src/v/bp_fe_icache.v]: that means those leaf modules do not handle zero width signals correctly and should be fixed

File: bp_common/src/include/bp_common_cache_engine_if.svh
Comment (dpetrisko) [inline bp_common/src/include/bp_common_cache_engine_if.svh:164]: nit, but do (safe_minus() + ways_p) to be explicit

File: bp_me/src/v/cce/bp_uce.v
Comment (dpetrisko) [reply bp_me/src/v/cce/bp_uce.v]: pretty much everywhere bank_offset_width_lp is used below, it is with BSG_SAFE_MINUS or BSG_MAX.  If it's set to 1 up here, then those will go away.  

Better to adjust in the parameter calculation than in the datapath, for clarity

File: bp_me/src/v/cce/bp_uce.v
Comment (dpetrisko) [reply bp_me/src/v/cce/bp_uce.v]: This comment is kind of out of place, doesn't really say much about the below code.  I'd combine it with the below comment or just get rid of it

File: bp_fe/src/v/bp_fe_icache.v
Comment (sripathi-muralitharan) [reply bp_fe/src/v/bp_fe_icache.v]: Fixed in basejump and opened a PR - https://github.com/bespoke-silicon-group/basejump_stl/pull/357

File: bp_common/src/include/bp_common_cache_engine_if.svh
Comment (sripathi-muralitharan) [reply bp_common/src/include/bp_common_cache_engine_if.svh]: Tried to do it the way that was suggested, but doing a safe_minus returns a 0 which is not very useful for width calculations. To make things more explicit, I added in a ?: to switch between ways = 1 and ways != 1.

Comment (dpetrisko) [reply bp_common/src/include/bp_common_cache_engine_if.svh]: how it was originally was fine

File: bp_me/src/v/cce/bp_uce.sv
Comment (dpetrisko) [inline bp_me/src/v/cce/bp_uce.sv:294]: // fill_index_shift corresponds to the single bank in a direct-mapped cache

File: bp_me/src/v/cce/bp_uce.sv
Comment (dpetrisko) [inline bp_me/src/v/cce/bp_uce.sv:24]: Nice! I think this module is minimally modified

File: bp_fe/src/v/bp_fe_icache.sv
Comment (dpetrisko) [inline bp_fe/src/v/bp_fe_icache.sv:700]: don't need

File: bp_fe/src/v/bp_fe_icache.sv
Comment (dpetrisko) [inline bp_fe/src/v/bp_fe_icache.sv:506]: I don't think you actually need this: a [-1:0] slice is perfectly legal, and with get removed below

File: bp_fe/src/v/bp_fe_icache.sv
Comment (dpetrisko) [inline bp_fe/src/v/bp_fe_icache.sv:380]: don't need conditional

File: bp_fe/src/v/bp_fe_icache.sv
Comment (dpetrisko) [inline bp_fe/src/v/bp_fe_icache.sv:121]: I also don't think you need the clause here since the signal is unused in direct mapped

File: bp_common/src/include/bp_common_cache_engine_if.svh
Comment (sripathi-muralitharan) [reply bp_common/src/include/bp_common_cache_engine_if.svh]: Added in a comment to remove future confusions and reverted back to the older version.

**Comment by sripathi-muralitharan** [comment]
In addition to the suggested comments, I also ended up removing a lot of the earlier changes from the I$ code. I think right now it is at the most minimally changed version.

File: bp_fe/src/v/bp_fe_icache.sv
Comment (dpetrisko) [inline bp_fe/src/v/bp_fe_icache.sv:176]: nice!

File: bp_me/src/v/cce/bp_uce.sv
Comment (dpetrisko) [inline bp_me/src/v/cce/bp_uce.sv:291]: Technically unrelated change, but I think it fits here. Can you change bank_sub_offset_width_lp to $clog2 instead of SAFE_CLOG2 and get rid of this if statement?

**Review by dpetrisko — CHANGES_REQUESTED** [review]
LGTM other than the last change requested and basejump going back to master once all PRs are merged.  Good work!

File: bp_me/src/v/cce/bp_uce.sv
Comment (sripathi-muralitharan) [reply bp_me/src/v/cce/bp_uce.sv]: Done.

File: bp_fe/src/v/bp_fe_icache.sv
Comment (sripathi-muralitharan) [reply bp_fe/src/v/bp_fe_icache.sv:176]: Thanks!

**Review by dpetrisko — APPROVED** [review]

**Pull request merged**
